src/Microsoft.Diagnostics.Monitoring.WebApi/CollectionRulePipelineState.cs

```diff
@@ -0,0 +1,238 @@
+                ActionCountSlidingWindowDuration = other.ActionCountSlidingWindowDuration;
+                RuleDuration = other.RuleDuration;
+                ExecutionTimestamps = new Queue<DateTime>(other.ExecutionTimestamps);
+                AllExecutionTimestamps = new List<DateTime>(other.AllExecutionTimestamps);
+                CurrentState = other.CurrentState;
+                CurrentStateReason = other.CurrentStateReason;
+            }
+        }
+
+        public CollectionRulePipelineState(int actionCountLimit, TimeSpan? actionCountSlidingWindowDuration, TimeSpan? ruleDuration)
+        {
+            ActionCountLimit = actionCountLimit;
+            ActionCountSlidingWindowDuration = actionCountSlidingWindowDuration;
+            RuleDuration = ruleDuration;
+            ExecutionTimestamps = new Queue<DateTime>(ActionCountLimit);
+            AllExecutionTimestamps = new List<DateTime>();
+            CurrentState = CollectionRuleState.Running;
+            CurrentStateReason = Strings.Message_CollectionRuleStateReason_Running;
+        }
+
+        public void BeginActionExecution()
+        {
+            lock (_lock)
+            {
+                Debug.Assert(CurrentState != CollectionRuleState.Finished);
+
+                CurrentState = CollectionRuleState.ActionExecuting;
+                CurrentStateReason = Strings.Message_CollectionRuleStateReason_ExecutingActions;
+            }
+        }
+
+        private void ActionExecutionSucceeded()
+        {
+            lock (_lock)
+            {
+                Debug.Assert(CurrentState == CollectionRuleState.ActionExecuting);
+
+                CurrentState = CollectionRuleState.Running;
+                CurrentStateReason = Strings.Message_CollectionRuleStateReason_Running;
+            }
+        }
+
+        private void ActionExecutionFailed()
+        {
+            lock (_lock)
+            {
+                Debug.Assert(CurrentState == CollectionRuleState.ActionExecuting);
+
+                CurrentState = CollectionRuleState.Running;
+                CurrentStateReason = Strings.Message_CollectionRuleStateReason_Running;
+            }
+        }
+
+        private void BeginThrottled()
+        {
+            lock (_lock)
+            {
+                Debug.Assert(CurrentState != CollectionRuleState.Finished);
+
+                CurrentState = CollectionRuleState.Throttled;
+                CurrentStateReason = Strings.Message_CollectionRuleStateReason_Throttled;
+            }
+        }
+
+        private void EndThrottled()
+        {
+            lock (_lock)
+            {
+                if (CurrentState == CollectionRuleState.Throttled)
+                {
+                    CurrentState = CollectionRuleState.Running;
+                    CurrentStateReason = Strings.Message_CollectionRuleStateReason_Running;
+                }
+            }
+        }
+
+        public void StartupTriggerCompleted()
```

Comment:
This, RuleDurationReached, ActionCountReached are the same thing but with different explanations. Can we merge these into one method with some parameters to determine which reason to specify?